### PR TITLE
feat: add docker compose and improve selfhosting

### PR DESCRIPTION
**Pull request by shafy**
This PR improves selfhosting and contributing by adding Docker and `docker-compose` setups for dev, prod and prod-remote (where the Mapzy Docker image is pulled from Docker Hub).

Futhermore, I've added slight changes to the code so that stuff like trial notifications, subscription signup in settings etc. don't show up for selfhoster, controlled by a new env variable `MAPZY_CLOUD`.

As requested by Elestio, there are three further env variables `INIT_USER_EMAIL`, `INIT_USER_PASSWORD` and `ALLOW_REGISTRATION`. If the first two are set, it creates a user in the database (see `docker-entrypoint.sh`). The latter disables signing up (useful if you're self-hosting).

I've also added a new GitHub Action that builds and pushes a Docker image everytime we createa a release in GitHub.

Oh, and OhMySMTP is now called MailPace, so I updated everything related to that.

Before deploying to DO, we need to update the build command (the prod Dockerfile is now called `Dockerfile.prod` just to be more explicit) and also add the new env vars.

Looking forward to your comments @sagzy, especially if you think how I handled the abstraction for clound-only code for the `User` with `Cloud::User::CloudEmails` makes sense (alternative would be a concern, but I think this makes it clearer, albeit more verbose).

### CODE REVIEW

**Review by sagzy — COMMENTED** [review]
Great stuff dude! Thanks a lot.

### Docker setups + self hosting guide
Fuckign  great. Added some minor comments about the wording.

### Cloud version vs self-hosting logic
I think in general the way we handle the difference is good enough for today. Here are some things I would do differently in the future:
- the multiple `return unless ENV[''] == x` in different model & service methods. This is not very DRY, but more importantly, it's prone to errors in the future (when adding a new method that would be cloud only, one has to remember to add this return line in the new method). If we compare this to `dashboard_controller.rb`, it's easier to the difference: whenever we need to add a feature to the dashboard, we just add it under the `/dashboard` folder and we know the common logic is already taken care of (e.g. all actions in the dashboard are authorized, we do not need to write `authenticate_user!` in every new method). Similarly, I think having the cloud vs self-hosting part handled like that in the controller(s) directly makes a lot of sense. Either with a parent controller class or a concern.

- email notification vs flash notifications: we now have two interesting examples in the code base on how not to send the notifications for self-hosters. The email notification is handled by the new `cloud_email` class and the flash notification uses a concern. Both have a way of sending the notification only when it's a cloud version. I personally prefer the flash notification version with a concern, as 1. it doesn't spam the main code, 2. the condition can be written only once at the top (we can return early from the entire concern if it's a self-hosting version) and 3. there is no need to instantiate a whole class to finally do nothing with it

### Allow registration and create an account automatically
This is interesting. I didn't think of creating new accounts as a problem for self-hosters, but it's a good point. I'm a bit sceptical about the user creation on the `docker_entrypoint.sh` with the bash script (especially the part that checks whether a user exists based on the console output). Shouldn't we have like a rake task instead that does it? And maybe let the self-hoster run the task manually themselves, similar to seeds, instead of bundling it with the docker setup?

File: SELFHOSTING.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# Self-Hosting Guide
+
+Mapzy is open-source and can be self-hosted without any feature restrictions. Follow this guide to get started with self-hosting. If you have any questions, don't hesitate asking on our [GitHub Discussions](https://github.com/mapzy/mapzy/discussions) page.
+
+## Environment variables
+
+Create an `.env` file in the root directory of the app. Use `.env.example` as a template. If you are running on a managed service like Heroku or DigitalOcean, enter the environment variables in their system.
+
+## Hosting
+Mapzy is a pure Ruby on Rails app requiring a PostgreSQL database (optionally also Redis). Therefore, you can host it however you like. Below, we've added short guides detailing a few popular self-hosting options.
```
[the remark this answers is not in the document]
Comment (shafy) [reply SELFHOSTING.md:10]: Yeah it's not needed. I guess they can add it if they want to get a welcome email or something :D But I can also remove this line.

File: docker-compose.dev.yml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      - REDIS_URL=redis://redis:6379
+      - MAPZY_CLOUD=${MAPZY_CLOUD}
+      - MAILPACE_TOKEN=${MAILPACE_TOKEN}
+      - STRIPE_ENDPOINT_SECRET=${STRIPE_ENDPOINT_SECRET}
+      - STRIPE_MAXI_PRICE_ID=${STRIPE_MAXI_PRICE_ID}
+      - STRIPE_MEDIUM_PRICE_ID=${STRIPE_MEDIUM_PRICE_ID}
+      - STRIPE_MINI_PRICE_ID=${STRIPE_MINI_PRICE_ID}
+      - STRIPE_SECRET_KEY=${STRIPE_SECRET_KEY}
+      - FUGU_API_KEY=${FUGU_API_KEY}
+      - FUGU_URL=${FUGU_URL}
```
[the remark this answers is not in the document]
Comment (shafy) [reply docker-compose.dev.yml:47]: I think they will look at `.env.example` anyways and there it's clear which are needed. Self-hosters won't use the this docker-compose file anyways, and contributors should be smart enough to see it from `.env.example`. I didn't want to duplicate the comments here.

> Do we actually need these env variables to be listed here again? Isn't it enough to pass the env file when starting docker compose (e.g. `docker-compose --env-file .env.dev up )

Yup, it's needed.

**Comment by shafy** [comment]
Thanks for your comments. Regarding DRYness:

> the multiple "return unless ENV[''] == x" in different model & service methods. This is not very DRY, but more importantly, it's prone to errors in the future (when adding a new method that would be cloud only, one has to remember to add this return line in the new method). If we compare this to dashboard_controller.rb, it's easier to the difference: whenever I need to add a feature to the dashboard, I just add it under the /dashboard folder and I know the common logic is already taken care of (e.g. all actions in the dashboard are authorized, I do not need to write authenticate_user! in every new method). Similarly, I think having the cloud vs self-hosting part handled like that in the controller(s) directly makes a lot of sense. Either with a parent controller class or a concern.

I think it's inevitable that we will have that guard clause with the env variable in multiple locations (e.g. template files, controllers and models).
I think I don't understand fully how this would work with inheritance similar to `DashboardController`. With `DashboardController`, we end up with two different controllers and actions, which makes sense, because we have different routes. But with the Cloud thing, we just want to disable one part of the controller, not have a separate controller and route.

> I personally prefer the flash notification version with a concern, as 1. it doesn't spam the main code, 2. the condition can be written only once at the top (we can return early from the entire concern if it's a self-hosting version) and 3. there is no need to instantiate a whole class to finally do nothing with it

In this case, we would have a concern with `ClassMethods`. So we still wound need to have the cloud guard clause for every method, we can't just do it like a `before_action` with the `TrialNotifiable` concern. I actually had it as a concern first, but then thought it's cleaner to abstract into the `cloud` folder.

Or maybe I'm not getting what you mean exactly. Could you make a commit where you would change the code like you suggest?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -1,16 +1,35 @@
+## Contribution Guide
+
+## Contribution Rules
+
+This is an open-source project and we welcome contributions that improve the code, UX/UI and add new features. However, this doesn't mean that we will accept all contributions. The best way to contribute is to check the open issues or ask us (bonjour@mapzy.io) if there's something you can contribute to.
```
[the remark this answers is not in the document]
Comment (shafy) [reply CONTRIBUTING.md:5]: Sure. Where/how?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -1,16 +1,35 @@
...
+Don't forget to create an `.env` file using `.env.example` as a template.
+
+If you're not using `docker-compose`, make sure to create the database with `docker run --env-file .env your_image rails db:create` before starting the Rails server.
+
+### Old school
 
 Dependencies:
 - ruby ~3.0.0
-- rails ~6.1.3
+- rails ~7.0.0
```
[the remark this answers is not in the document]
Comment (shafy) [reply CONTRIBUTING.md:32]: Oops, I copy pasted this from Fugu where I already have Rails 7. FALSE ALARM.

File: docker-compose.dev.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (shafy) [reply docker-compose.dev.yml:47]: Altough, thinking about it, maybe not. Let me try with passing the env file explicitly, would be much cleaner if it works.

[the remark this answers is not in the document]
Comment (shafy) [reply docker-compose.dev.yml:47]: @sagzy I double checked this, and indeed, if we pass an `.env` file explicitly, we don't need to list the env variables in the compose file. BUT: This requires the hoster to define and `.env` file, and it doesn't work with env variables defined in the shell without an `.env` file. Explicitly listing the env vars in the compose files works with both (`.env` file or env variables defined in the shell). Therefore, I would keep it like it is.

File: CONTRIBUTING.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (sagzy) [reply CONTRIBUTING.md:5]: Github Discussion directly

**Comment by shafy** [comment]
@sagzy I'm trying to implement what we discussed offline last week. But I think it won't work with the concerns as we thought. For example here:

```ruby
module Cloud
  module RegistrationWorkable
    extend ActiveSupport::Concern

    include Sidekiq::Worker
    
    # We can't do the following, because it's not an action, but a controller method

    # included do
    #   before_action :return_early, unless: :mapzy_cloud?, only: [:setup_registration_workers]
    # end
   
    # Instead, we need to return early in each method
    def send_welcome_email(email)
      return unless mapzy_cloud?

      AccountMailer.with(email: email).welcome_email.deliver_later
    end

    def setup_registration_workers(user_id)
      return unless mapzy_cloud?

      EmailWorker.perform_at(7.days.from_now, "reminder_email1", user_id)
      EmailWorker.perform_at(13.days.from_now, "reminder_email2", user_id)
      EmailWorker.perform_at(14.days.from_now, "account_inactivated_email", user_id)
      AccountWorker.perform_at(14.days.from_now, user_id)
    end
  end
end
```

I don't think we'll get around returning early in each method without going for a more fancy and complicated solution (see comments above). Am I missing something?

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -1,16 +1,35 @@
...
 
-We recommend developing with the excellent cloud-based development environment provided by [Gitpod](gitpod.io/). To start your ready-to-code development enviroment just add `gitpod.io/#` in front of the GitHub URL for any branch, or [click here](https://gitpod.io/#https://github.com/mapzy/mapzy) to get started with the main branch.
+### Docker
 
-Run `foreman start -f Procfile.dev` to start the Rails server and the Webpack Dev Server (or run the processes in separate terminals manually).
+You can build the Docker image using the `Dockerfile.dev` configuration. If you don't have an existing PostgreSQL service on your machine, you can use our `docker-compose` setup to also start a PostgreSQL database and Redis. Just run `docker-compose -f docker-compose.dev.yml up`.
```
[the remark this answers is not in the document]
Comment (shafy) [reply CONTRIBUTING.md]: Fixed in 45ba98e

File: CONTRIBUTING.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (shafy) [reply CONTRIBUTING.md]: Fixed in 45ba98e

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -21,9 +40,13 @@ cp .env.example .env
...
-Run `foreman start -f Procfile.dev` to start the Rails server and the Webpack Dev Server (or run the processes in separate terminals manually).
\ No newline at end of file
+## Starting the development server
+To start the server, run `.foreman start -f Procfile.dev`. This starts the Rails server, the Tailwind watch process and the Sidekiq process.
+
+
+## Mapzy Cloud
+There are a few places in the codebase where you will encounter logic that is only needed for Mapzy Cloud, the hosted solution by us. These parts of the code are not needed if you are self-hosting, and probably also not relevant for you as a contributor. For the most part this concerns code that relates to payment and subscriptions.
+
+Currently, we use the environment variable `MAPZY_CLOUD` to determine if the instance is being self-hosted or not. Later on, we might improve this structure by factoring out the main code base to a Rails Engine.
```
[the remark this answers is not in the document]
Comment (shafy) [reply CONTRIBUTING.md]: Fixed in 45ba98e

File: SELFHOSTING.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shafy) [reply SELFHOSTING.md]: Fixed in 45ba98e

File: app/views/devise/registrations/new.html.erb [hunk 7]
```diff
@@ -1,47 +1,57 @@
 <%= render "devise/shared/layout" do %>
   <h1 class="text-center mb-8 md:mb-12">Create your store locator</h1>
 
-  <%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
-    <div class="field mt-4">
-      <%= f.label :email, class: "label--default" %>
-      <%= f.email_field :email, autofocus: true, autocomplete: "email", required: true,
-        class: "input--default" %>
-    </div>
+  hello
```
[the remark this answers is not in the document]
Comment (shafy) [reply app/views/devise/registrations/new.html.erb]: Fixed in 45ba98e

File: app/views/devise/registrations/new.html.erb [hunk 8]
```diff
@@ -1,47 +1,57 @@
...
+  <%= ENV["ALLOW_REGISTRATION"] %>
```
[the remark this answers is not in the document]
Comment (shafy) [reply app/views/devise/registrations/new.html.erb]: Removed in 45ba98e

**Comment by shafy** [comment]
@sagzy made the changes as discussed above. Please have a quick look if there's a big no-no so we can finish this PR.

**Comment by sagzy** [comment]
Thanks @shafy for the changes. Yes I agree, it's pretty clean and readable with your solution, as all the cloud code is nested in one place. I added one comment for one the Stripe concern, which I think may break.

> But I think it won't work with the concerns as we thought. I don't think we'll get around returning early in each method without going for a more fancy and complicated solution

Yes I agree, I mean it works fine if we have separate routes for the cloud vs non-cloud controllers, as then we can return early from the entire controller. For example, the StripeController would be a good place to implement something like that:

```
module Cloudable
  before_action :verify_cloud
 
  def verify_cloud
     head :not_found if mapzy_cloud?
  end
end
```

```
class StripeController < ApplicationController
   include Cloudable
   
   # We have returned early from all the controller already
   ...
end
```

but for the rest of the cloud vs non-cloud functionality (like the flash notifications, the email notifications, the account setting view), it would require us to duplicate all the dashboard controllers for little upside. So I like your solution better. 👍

File: app/controllers/concerns/cloud/stripeable.rb [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    def stripe_product
+      @stripe_product ||= stripe_subscription.items&.data&.first&.price&.product
+    end
+
+    def stripe_subscription
+      @stripe_subscription ||= stripe_customer.subscriptions&.data&.first
+    end
+
+    def stripe_customer
+      return unless mapzy_cloud?
```
Comment (sagzy) [inline app/controllers/concerns/cloud/stripeable.rb:18]:
```
    def stripe_customer
      return unless mapzy_cloud?
      ...
     end
```

as `stripe_customer` can be nil, we should be more careful in the other methods calling it (`stripe_subscription` and `stripe_product`).

File: app/helpers/cloud_helper.rb [hunk 10]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+module CloudHelper
+  def mapzy_cloud?
+    ENV["MAPZY_CLOUD"] == "true"
+  end
+end
```
Comment (sagzy) [inline app/helpers/cloud_helper.rb:7]: Both `allow_registration?` and `mapzy_cloud?` can be in ApplicationHelper directly I think

File: app/controllers/concerns/cloud/stripeable.rb (hunk 9 above)
Comment (shafy) [reply app/controllers/concerns/cloud/stripeable.rb:18]: True, but in the current case the other methods never call `stripe_customer` if it's `blank`. I guess should be good for now?

File: app/helpers/cloud_helper.rb (hunk 10 above)
Comment (shafy) [reply app/helpers/cloud_helper.rb]: Fixed in ef8a446

**Pull request merged**